mplc/multi_partner_learning.py

```diff
@@ -663,6 +664,724 @@ def fit_minibatch(self):
+        self.partners_weights = [[tf.Variable(initial_value=w.read_value()) for w in
+                                  self.model.trainable_weights] for _ in self.partners_list]
+        self.partners_optimizers = [self.model.optimizer.from_config(self.model.optimizer.get_config()) for _ in
+                                    self.partners_list]
+
+    def log_epoch(self, epoch_number, history):
+        logger.info(
+            f'[{self.name}] > Epoch {str(epoch_number + 1).ljust(2)}/{self.epoch_count} -'
+            f' {f"{np.round(time.time() - self.epoch_timer)} s.".ljust(6)} >'
+            f' {" -- ".join(f"{key}: {str(np.round(value, 2)).ljust(5)}" for key, value in history.items())}')
+        if not self.history:
+            self.history = {key: [value] for key, value in history.items()}
+        else:
+            for key, value in history.items():
+                self.history[key].append(value)
+
+    def log_end_training(self):
+        training_time = (time.time() - self.timer)
+        self.end_test_score = self.model.evaluate(self.test_data, return_dict=True, verbose=False)
+        logger.info(
+            f'[{self.name}] > Training {self.epochs_index} epoch in {np.round(training_time, 3)} seconds.'
+            f' Tests scores: '
+            f'{" -- ".join(f"{key}: {np.around(value, 3)}" for key, value in self.end_test_score.items())}')
+        self.learning_computation_time += training_time
+
+    def get_epoch_history(self):
+        # get the metrics computed. Here the metrics are computed on all the registered state (y, y_pred) stored,
+        # which means on all the minibatches, for all partners.
+        history = {m.name: float((m.result().numpy())) for m in self.model.metrics}
+        # compute val accuracy and loss for global model
+        val_hist = self.model.evaluate(self.val_data, return_dict=True, verbose=False)
+        history.update({f'val_{key}': value for key, value in val_hist.items()})
+        self.model.reset_metrics()
+        return history
+
+    def early_stop(self):
+        """
+        If the metric monitored decreases (or increases, depends on the metrics monitored)
+        during patience epoch, we stop the learning.
+        """
+        if self.monitor_op(self.history[self.monitor_value][-1] - self.min_delta, self.best):
+            self.wait = 0
+            self.best = self.history[self.monitor_value][-1]
+        else:
+            self.wait += 1
+            if self.wait >= self.patience:
+                return True
+        return False
+
+    def fit(self):
+
+        # TF function definition
+        @tf.function
+        def aggregate_weights(weights, agg_w):
+            res = list()
+            for weights_per_layer in zip(*weights):
+                res.append(tf.tensordot(weights_per_layer, agg_w, [0, 0]))
+            return res
+
+        @tf.function
+        def fit_minibatch(model, partners_minibatches, partners_optimizers, partners_weights, agg_w):
```

Comment:
in this function we fit several minibatches?
If so we should probably rename it to fit_minibatches()?

Reply:
In fact we perform one minibatch for each partner.